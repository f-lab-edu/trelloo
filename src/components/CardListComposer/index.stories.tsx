import React from "react";
import CardListComposer from "@components/CardListComposer";
import { AddListRequest } from "@/queries/cards/interface";

export default {
  title: "Components/forms",
  component: CardListComposer,
};

const Template = () => {
  const handleAddList = async ({ title }: AddListRequest) => {};

  return <CardListComposer onAddList={handleAddList} />;
};

export const CardListComposerComponent = Template.bind({});
