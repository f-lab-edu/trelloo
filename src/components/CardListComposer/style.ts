import styled from "styled-components";

export const Container = styled.div`
  margin-left: 10px;
  padding: 8px;
  min-width: 272px;
  background-color: ${({ theme }) => theme.color.cardListGray};
  border-radius: ${({ theme }) => theme.borderRadius.card};
  height: 90px;
`;
