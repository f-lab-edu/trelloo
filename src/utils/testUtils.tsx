import { store } from "@/store";
import { GlobalStyle } from "@/styles/GlobalStyle";
import { theme } from "@/styles/theme";
import ModalsProvider from "@components/modals/ModalsProvider";
import { QueryClient, QueryClientProvider, UseQueryResult } from "@tanstack/react-query";
import { render, renderHook, RenderOptions } from "@testing-library/react";
import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import { ThemeProvider } from "styled-components";

const TestProviders = ({ children }: { children: React.ReactNode }) => {
  const queryClient = new QueryClient();

  return (
    <QueryClientProvider client={queryClient}>
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <GlobalStyle />
          <ModalsProvider>{children}</ModalsProvider>
        </ThemeProvider>
      </Provider>
    </QueryClientProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: TestProviders, ...options });

const createQueryClientWrapper = ({ children }: { children: React.ReactNode }) => {
  const queryClient = new QueryClient();

  return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>;
};

export const createMockedQuery = (hook: () => UseQueryResult<any, unknown>) => {
  return renderHook(() => hook(), { wrapper: createQueryClientWrapper });
};

export * from "@testing-library/react";
export { customRender };
