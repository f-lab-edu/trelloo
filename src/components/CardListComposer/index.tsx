import React, { useState } from "react";
import { Input } from "antd";
import { AddListRequest } from "@/queries/cards/interface";
import Composer from "@components/forms/Composer";
import { useController, useForm } from "react-hook-form";
import * as S from "./style";

interface Props {
  onAddList: (params: AddListRequest) => Promise<void>;
}

function CardListComposer({ onAddList }: Props) {
  const [isInputOpened, setIsInputOpened] = useState(false);

  const handleInputOpen = () => {
    setIsInputOpened(!isInputOpened);
  };

  const { control, reset, handleSubmit } = useForm({
    defaultValues: { title: "" },
    mode: "onSubmit",
  });

  const {
    field: { onChange, value },
  } = useController({
    name: "title",
    control,
    rules: { required: true },
  });

  const handleAddList = async () => {
    await onAddList({
      title: value,
    });
    reset();
  };

  const wrappedOnSubmit = handleSubmit(handleAddList);

  return (
    <S.Container>
      <Composer
        isOpen={isInputOpened}
        toggleInputOpen={handleInputOpen}
        btnText="Add a list"
        submitBtnText="Add list"
        onSubmit={wrappedOnSubmit}
      >
        <Input onChange={onChange} value={value} placeholder="Enter list title..." />
      </Composer>
    </S.Container>
  );
}

export default CardListComposer;
