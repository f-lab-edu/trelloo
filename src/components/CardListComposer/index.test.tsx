import React from "react";
import { customRender, screen } from "@utils/testUtils";
import { beforeEach, describe, vi } from "vitest";
import CardListComposer from ".";

describe("CardListComposer 테스트", () => {
  const handleAddList = vi.fn();

  beforeEach(() => {
    customRender(<CardListComposer onAddList={handleAddList} />);
  });

  it("컴포넌트 렌더링", () => {
    expect(screen.getByRole("button")).toHaveTextContent("Add a list");
  });

  it("스냅샷 테스트", () => {
    const component = customRender(<CardListComposer onAddList={handleAddList} />);
    expect(component).toMatchSnapshot();
  });
});
