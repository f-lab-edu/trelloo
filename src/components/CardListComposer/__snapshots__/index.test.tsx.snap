// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardListComposer 테스트 > 스냅샷 테스트 1`] = `
{
  "getByPlaceholderText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
}
`;
